[View in Colaboratory](https://colab.research.google.com/github/KHATUNA3/Week2_public/blob/master/09_Conv1Ds.ipynb)

# 1d Convolutional Neural Networks

In [0]:

# Imports
import numpy as np
import os
# from utils.utilities import *
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import tensorflow as tf
import pandas as pd
import tensorflow as tf
from tensorflow import keras

### About the Data

Data is from people doing one of these activities (WALKING, WALKING_UPSTAIRS, WALKING_DOWNSTAIRS, SITTING, STANDING, LAYING) <br><br>
The data is from a smartphone and provides the body acceleration, general acceleration and a gyroscope reading <br><br>
Data is from: https://archive.ics.uci.edu/ml/datasets/human+activity+recognition+using+smartphones 

### Read in data from functions provided below

In [0]:
X_train, labels_train, list_ch_train = read_data_train() # train
X_test, labels_test, list_ch_test = read_data_test()

In [47]:
print(X_train.shape)

(7352, 128, 9)



#### Explaining the shape of the data 
There are 7,352 data points that are classified as doing one of those activities <br>
There are 128 time steps <br>
There are 9 values for each time step for each data point for the x,y,z values of the body 
acceleration, general acceleration and gyrscope reading. <br>

### Use train_test_split to create training and validation data and labels

In [0]:
# your code here
X_tr, X_val, lb_tr, lb_val = train_test_split(X_train, labels_train, stratify = labels_train, random_state = 123)

One hot coding is when each row represents one label and it the nonzero column represents what the label is<br>
 [0,0,0,0,0,1] <br>
           ^ This label is LAYING, which was previously represented with a 5
           <br><br>

### One hot encoding functions are provided at the bottom of this notebook <br> Use these functions to convert the training, validation and testing labels

In [0]:
# your code here
y_tr = one_hot(lb_tr)
y_val = one_hot(lb_val)
y_test = one_hot(labels_test)

### Define your Keras model <br>
You wlil want to use

See for examples on creating Keras modelshttps://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py

In [50]:
# fill in with your code below
#from keras.layers import Conv1D, MaxPooling2D, Dropout

model = keras.Sequential()
model.add(keras.layers.Conv1D(filters=18, kernel_size=3, strides=1, padding='same', activation='relu', input_shape=(128,9)))
model.add(keras.layers.GlobalMaxPooling1D())
model.add(keras.layers.Dense(6, activation = tf.nn.sigmoid))
print(model.summary())  # view model
    
    


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d_17 (Conv1D)           (None, 128, 18)           504       
_________________________________________________________________
global_max_pooling1d_3 (Glob (None, 18)                0         
_________________________________________________________________
dense_12 (Dense)             (None, 6)                 114       
Total params: 618
Trainable params: 618
Non-trainable params: 0
_________________________________________________________________
None


### Compile your model

In [0]:
lr = 0.0008 # choose a learning rate, this will be a good paramter to tune
model.compile( optimizer=tf.train.AdamOptimizer(lr), loss=keras.losses.categorical_crossentropy, metrics = ['accuracy'] )

### Fit your model

In [52]:
history = model.fit( X_tr,y_tr,
                    epochs= 100,
                    batch_size= 600,
                    validation_data=(X_val, y_val),
                    verbose=1 
                    )

Train on 5514 samples, validate on 1838 samples
Epoch 1/100
5514/5514 [==============================] - 2s 282us/step - loss: 1.8286 - acc: 0.2994 - val_loss: 1.7962 - val_acc: 0.3188
Epoch 2/100
5514/5514 [==============================] - 1s 207us/step - loss: 1.7775 - acc: 0.3172 - val_loss: 1.7516 - val_acc: 0.3264
Epoch 3/100
5514/5514 [==============================] - 1s 240us/step - loss: 1.7362 - acc: 0.3241 - val_loss: 1.7130 - val_acc: 0.3379
Epoch 4/100
5514/5514 [==============================] - 1s 239us/step - loss: 1.6998 - acc: 0.3411 - val_loss: 1.6779 - val_acc: 0.3651
Epoch 5/100
5514/5514 [==============================] - 1s 239us/step - loss: 1.6654 - acc: 0.3672 - val_loss: 1.6438 - val_acc: 0.3885
Epoch 6/100
5514/5514 [==============================] - 1s 236us/step - loss: 1.6316 - acc: 0.3856 - val_loss: 1.6098 - val_acc: 0.3993
Epoch 7/100
5514/5514 [==============================] - 1s 241us/step - loss: 1.5981 - acc: 0.4006 - val_loss: 1.5765 - val_acc: 

Epoch 12/100
5514/5514 [==============================] - 1s 231us/step - loss: 1.4480 - acc: 0.4458 - val_loss: 1.4274 - val_acc: 0.4652
Epoch 13/100
5514/5514 [==============================] - 1s 230us/step - loss: 1.4195 - acc: 0.4946 - val_loss: 1.3975 - val_acc: 0.5435
Epoch 14/100
5514/5514 [==============================] - 1s 229us/step - loss: 1.3892 - acc: 0.5555 - val_loss: 1.3662 - val_acc: 0.5838
Epoch 15/100
5514/5514 [==============================] - 1s 230us/step - loss: 1.3579 - acc: 0.5878 - val_loss: 1.3343 - val_acc: 0.6257
Epoch 16/100
5514/5514 [==============================] - 1s 231us/step - loss: 1.3259 - acc: 0.6503 - val_loss: 1.3012 - val_acc: 0.6942
Epoch 17/100
5514/5514 [==============================] - 1s 230us/step - loss: 1.2928 - acc: 0.7035 - val_loss: 1.2673 - val_acc: 0.7274
Epoch 18/100
5514/5514 [==============================] - 1s 230us/step - loss: 1.2587 - acc: 0.7372 - val_loss: 1.2322 - val_acc: 0.7541
Epoch 19/100
5514/5514 [==========

5514/5514 [==============================] - 1s 228us/step - loss: 1.0750 - acc: 0.8058 - val_loss: 1.0458 - val_acc: 0.8243
Epoch 24/100
5514/5514 [==============================] - 1s 229us/step - loss: 1.0372 - acc: 0.8166 - val_loss: 1.0076 - val_acc: 0.8264
Epoch 25/100
5514/5514 [==============================] - 1s 232us/step - loss: 0.9994 - acc: 0.8250 - val_loss: 0.9699 - val_acc: 0.8319
Epoch 26/100
5514/5514 [==============================] - 1s 229us/step - loss: 0.9620 - acc: 0.8332 - val_loss: 0.9328 - val_acc: 0.8466
Epoch 27/100
5514/5514 [==============================] - 1s 229us/step - loss: 0.9253 - acc: 0.8415 - val_loss: 0.8967 - val_acc: 0.8526
Epoch 28/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.8891 - acc: 0.8487 - val_loss: 0.8610 - val_acc: 0.8553
Epoch 29/100
5514/5514 [==============================] - 1s 229us/step - loss: 0.8540 - acc: 0.8538 - val_loss: 0.8265 - val_acc: 0.8624
Epoch 30/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 230us/step - loss: 0.6988 - acc: 0.8818 - val_loss: 0.6768 - val_acc: 0.8928
Epoch 35/100
5514/5514 [==============================] - 1s 231us/step - loss: 0.6724 - acc: 0.8865 - val_loss: 0.6514 - val_acc: 0.8934
Epoch 36/100
5514/5514 [==============================] - 1s 229us/step - loss: 0.6475 - acc: 0.8901 - val_loss: 0.6277 - val_acc: 0.8934
Epoch 37/100
5514/5514 [==============================] - 1s 229us/step - loss: 0.6242 - acc: 0.8917 - val_loss: 0.6051 - val_acc: 0.8972
Epoch 38/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.6027 - acc: 0.8939 - val_loss: 0.5842 - val_acc: 0.8993
Epoch 39/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.5821 - acc: 0.8983 - val_loss: 0.5651 - val_acc: 0.9032
Epoch 40/100
5514/5514 [==============================] - 1s 231us/step - loss: 0.5633 - acc: 0.9006 - val_loss: 0.5464 - val_acc: 0.9059
Epoch 41/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 232us/step - loss: 0.4839 - acc: 0.9066 - val_loss: 0.4706 - val_acc: 0.9081
Epoch 46/100
5514/5514 [==============================] - 1s 231us/step - loss: 0.4709 - acc: 0.9081 - val_loss: 0.4583 - val_acc: 0.9097
Epoch 47/100
5514/5514 [==============================] - 1s 232us/step - loss: 0.4587 - acc: 0.9091 - val_loss: 0.4460 - val_acc: 0.9108
Epoch 48/100
5514/5514 [==============================] - 1s 228us/step - loss: 0.4469 - acc: 0.9093 - val_loss: 0.4351 - val_acc: 0.9124
Epoch 49/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.4362 - acc: 0.9104 - val_loss: 0.4248 - val_acc: 0.9151
Epoch 50/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.4257 - acc: 0.9115 - val_loss: 0.4145 - val_acc: 0.9146
Epoch 51/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.4160 - acc: 0.9126 - val_loss: 0.4050 - val_acc: 0.9162
Epoch 52/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 228us/step - loss: 0.3747 - acc: 0.9148 - val_loss: 0.3652 - val_acc: 0.9157
Epoch 57/100
5514/5514 [==============================] - 1s 233us/step - loss: 0.3675 - acc: 0.9166 - val_loss: 0.3582 - val_acc: 0.9162
Epoch 58/100
5514/5514 [==============================] - 1s 231us/step - loss: 0.3608 - acc: 0.9168 - val_loss: 0.3519 - val_acc: 0.9184
Epoch 59/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.3547 - acc: 0.9159 - val_loss: 0.3453 - val_acc: 0.9162
Epoch 60/100
5514/5514 [==============================] - 1s 231us/step - loss: 0.3482 - acc: 0.9169 - val_loss: 0.3397 - val_acc: 0.9173
Epoch 61/100
5514/5514 [==============================] - 1s 232us/step - loss: 0.3421 - acc: 0.9175 - val_loss: 0.3336 - val_acc: 0.9206
Epoch 62/100
5514/5514 [==============================] - 1s 232us/step - loss: 0.3364 - acc: 0.9175 - val_loss: 0.3282 - val_acc: 0.9189
Epoch 63/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 229us/step - loss: 0.3114 - acc: 0.9236 - val_loss: 0.3040 - val_acc: 0.9222
Epoch 68/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.3069 - acc: 0.9233 - val_loss: 0.3000 - val_acc: 0.9211
Epoch 69/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.3026 - acc: 0.9238 - val_loss: 0.2956 - val_acc: 0.9222
Epoch 70/100
5514/5514 [==============================] - 1s 229us/step - loss: 0.2985 - acc: 0.9244 - val_loss: 0.2916 - val_acc: 0.9238
Epoch 71/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.2947 - acc: 0.9256 - val_loss: 0.2874 - val_acc: 0.9227
Epoch 72/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.2908 - acc: 0.9258 - val_loss: 0.2837 - val_acc: 0.9222
Epoch 73/100
5514/5514 [==============================] - 1s 231us/step - loss: 0.2871 - acc: 0.9246 - val_loss: 0.2806 - val_acc: 0.9227
Epoch 74/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 229us/step - loss: 0.2704 - acc: 0.9256 - val_loss: 0.2641 - val_acc: 0.9233
Epoch 79/100
5514/5514 [==============================] - 1s 229us/step - loss: 0.2673 - acc: 0.9256 - val_loss: 0.2611 - val_acc: 0.9249
Epoch 80/100
5514/5514 [==============================] - 1s 231us/step - loss: 0.2644 - acc: 0.9253 - val_loss: 0.2584 - val_acc: 0.9249
Epoch 81/100
5514/5514 [==============================] - 1s 231us/step - loss: 0.2615 - acc: 0.9260 - val_loss: 0.2556 - val_acc: 0.9227
Epoch 82/100
5514/5514 [==============================] - 1s 227us/step - loss: 0.2589 - acc: 0.9260 - val_loss: 0.2527 - val_acc: 0.9233
Epoch 83/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.2561 - acc: 0.9267 - val_loss: 0.2502 - val_acc: 0.9244
Epoch 84/100
5514/5514 [==============================] - 1s 229us/step - loss: 0.2535 - acc: 0.9267 - val_loss: 0.2480 - val_acc: 0.9260
Epoch 85/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 232us/step - loss: 0.2416 - acc: 0.9273 - val_loss: 0.2363 - val_acc: 0.9271
Epoch 90/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.2394 - acc: 0.9273 - val_loss: 0.2338 - val_acc: 0.9260
Epoch 91/100
5514/5514 [==============================] - 1s 232us/step - loss: 0.2372 - acc: 0.9276 - val_loss: 0.2318 - val_acc: 0.9271
Epoch 92/100
5514/5514 [==============================] - 1s 233us/step - loss: 0.2352 - acc: 0.9276 - val_loss: 0.2302 - val_acc: 0.9287
Epoch 93/100
5514/5514 [==============================] - 1s 233us/step - loss: 0.2332 - acc: 0.9280 - val_loss: 0.2279 - val_acc: 0.9276
Epoch 94/100
5514/5514 [==============================] - 1s 230us/step - loss: 0.2312 - acc: 0.9285 - val_loss: 0.2260 - val_acc: 0.9271
Epoch 95/100
5514/5514 [==============================] - 1s 228us/step - loss: 0.2291 - acc: 0.9284 - val_loss: 0.2243 - val_acc: 0.9271
Epoch 96/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 229us/step - loss: 0.2202 - acc: 0.9300 - val_loss: 0.2155 - val_acc: 0.9282


### Code below prints the test loss and accuracy <br>
### Before you print your accuracy and loss, play around with the parameters!!

In [53]:
score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

('Test loss:', 0.3847692847140652)
('Test accuracy:', 0.8629114353579912)


### Provided Functions Below

In [0]:
def read_data_train():
  """ Read data """

  # Fixed params
  n_class = 6
  n_steps = 128

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/y_train.txt'
  labels = pd.read_csv(label_path, header = None)

  list_of_channels = ['body_acc_x', 'body_acc_y', 'body_acc_z', 'body_gyro_x',
  'body_gyro_y', 'body_gyro_z', 'total_acc_x', 'total_acc_y', 'total_acc_z']

  X = np.zeros((len(labels), n_steps, len(list_of_channels)))

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,0] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,1] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,2] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,3] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,4] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,5] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,6] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,7] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,8] = dat_.as_matrix()


  # Return 
  return X, labels[0].values, list_of_channels

In [0]:
def read_data_test():
  """ Read data """

  # Fixed params
  n_class = 6
  n_steps = 128

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/y_test.txt'
  labels = pd.read_csv(label_path, header = None)

  list_of_channels = ['body_acc_x', 'body_acc_y', 'body_acc_z', 'body_gyro_x',
  'body_gyro_y', 'body_gyro_z', 'total_acc_x', 'total_acc_y', 'total_acc_z']

  X = np.zeros((len(labels), n_steps, len(list_of_channels)))

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,0] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,1] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,2] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,3] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,4] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,5] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,6] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,7] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,8] = dat_.as_matrix()


  # Return 
  return X, labels[0].values, list_of_channels

In [0]:
def one_hot(labels, n_class = 6):
	""" One-hot encoding """
	expansion = np.eye(n_class)
	y = expansion[:, labels-1].T
	assert y.shape[1] == n_class, "Wrong number of labels!"

	return y

In [0]:
def get_batches(X, y, batch_size = 100):
	""" Return a generator for batches """
	n_batches = len(X) // batch_size
	X, y = X[:n_batches*batch_size], y[:n_batches*batch_size]

	# Loop over batches and yield
	for b in range(0, len(X), batch_size):
		yield X[b:b+batch_size], y[b:b+batch_size]